### PR TITLE
Integ tests

### PR DESCRIPTION
**Pull request by sharad-oss**
*Issue #, if available:*
https://github.com/awslabs/amazon-emr-user-role-mapper/issues/5

*Description of changes:*
- Ability to run integration tests as part of `mvn clean install`
- Tests only run on OSX and Unix style OS, as we need the notion of userId, groupId etc. This might be changed in future.
- On non supported OS the build will succeed with no integration tests, but unit tests will still run
- All tests work by hitting the localhost server (without DNAT rules).
- The user running the tests is determined, and his userid is used to determine the caller.
- All other functionality is the same as the regular URM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java [hunk 1]
```diff
@@ -52,10 +52,10 @@
     @PostConstruct
     void init() {
         try {
-            String className = applicationConfiguration.getProperty(Constants.ROLE_MAPPER_CLASS,
+          String className = applicationConfiguration.getProperty(Constants.ROLE_MAPPER_CLASS,
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java:55]: 4 spaces?

File: emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java [hunk 2]
```diff
@@ -70,11 +70,11 @@ void init() {
             roleMapperProvider.init(applicationConfiguration.asMap());
             log.info("Initialized the mapper.");
         } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
-            log.error("Could not load the mapper", e);
+            log.info("Could not load the mapper " + e.getMessage());
             throw new RuntimeException("Could not load the mapper class", e);
         } catch (Throwable t) {
-            log.error("Could not load the mapper", t);
-            throw new RuntimeException("Could not initialize the mapper", t);
+          log.info("Could not load the mapper " + t.getMessage());
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java:76]: why info?

File: emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java [hunk 3]
```diff
@@ -70,11 +70,11 @@ void init() {
             roleMapperProvider.init(applicationConfiguration.asMap());
             log.info("Initialized the mapper.");
         } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
-            log.error("Could not load the mapper", e);
+            log.info("Could not load the mapper " + e.getMessage());
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java:73]: why info?

File: emr-user-role-mapper-application/pom.xml [hunk 4]
```diff
@@ -106,7 +106,12 @@
             <artifactId>aws-java-sdk-sts</artifactId>
             <version>${aws-java-sdk.version}</version>
         </dependency>
-        <dependency>
+      <dependency>
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/pom.xml:109]: looks like the indentation is incorrect. pls fix

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestBase.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      .setPrettyPrinting()
+      .create();
+  public static int RELOAD_CFG_TIME_MIN = 1;
+  public static String DEFAULT_MAPPER_IMPL_BUCKET = "urm-integ-test-default-mapper";
+  public static String DEFAULT_MAPPER_IMPL_MAPPING = "default-impl.json";
+  protected static String LOCALHOST_SERVER = "http://localhost";
+  protected static String IMDS_CREDENTIALS_URI = "/latest/meta-data/iam/security-credentials/";
+  protected static String USER_TO_CHANGE = "USER_TO_CHANGE";
+  protected static String USER_ROLE_TO_CHANGE = "USER_ROLE_TO_CHANGE";
+  protected static String AWS_ACCOUNT_TO_CHANGE = "AWS_ACCOUNT_TO_CHANGE";
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestBase.java:23]: Let's make the value as placeholder AWS_ACCOUNT_TO_CHANGE = "{AWS_ACCOUNT_TO_CHANGE}"? Similar to other things that will be substituted.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestsUserService.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class IntegrationTestsUserService implements UserIdService {
+
+  @Override
+  public OptionalInt resolveSystemUID(String localAddr, int localPort, String remoteAddr,
+      int remotePort, boolean isNativeIMDSApi) {
+     Optional<String> uid = new CommandBasedPrincipalResolver()
+        .runCommand(Arrays.asList("id", "-u")).stream().findFirst();
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestsUserService.java:17]: Instead of making runCommand public, can we just create a test class that does same thing?

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+    log.debug("Starting the Jetty server");
+    jettyServer.start();
+    // AWS Role creation is eventually consistent.
+    // TODO - Fine tune this.
+    Thread.sleep(10 * 1000);
+  }
+
+  @AfterClass
+  public static void tearDown() throws Exception {
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java:117]: don't we need to clean up the uploaded s3 contents as well?

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+    AmazonS3 s3 = AmazonS3ClientBuilder.standard()
+        .withCredentials(new AWSStaticCredentialsProvider(sessionCredentials))
+        .build();
+    S3Object s3Object = s3.getObject(new GetObjectRequest(TEST_CFG_BUCKET, TEST_CFG_OBJECT));
+    assertThat(S3Utils.getS3FileAsString(s3Object), is(TEST_S3_OBJECT_CONTENTS));
+  }
+
+  @Test
+  public void reload_config_no_mapping_no_credentials() throws Exception {
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java:174]: I think this test is trying to test on the case where URM periodically syncs the mapping file from s3. Should we rename this file to be more explicit? Especially, this test does not verify anything around credential or empty mapping.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    return String.format("arn:aws:iam::%s:policy/%s", account, policyName);
+  }
+
+  public static void attachPolicyToRole(String roleArn, String PolicyArn) {
+    AttachRolePolicyRequest attach_request =
+        new AttachRolePolicyRequest()
+            .withRoleName(roleArn)
+            .withPolicyArn(PolicyArn);
+
+    iam.attachRolePolicy(attach_request);
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java:53]: What happens if the policy is already attached?

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * Utility methods for AWS IAM.
+ */
+public class IAMUtils {
+
+  final static AmazonIdentityManagement iam =
+      AmazonIdentityManagementClientBuilder.defaultClient();
+
+
+  public static Policy createPolicy(String account, String policy) {
```
Comment (yInnovation12) [inline emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java:28]: it looks like for any logged user, only one integ test run can live at a given time. If multiple users logged in as the same user, then test runs may interfere with each other given the policies and mapping file would be exactly the same. This may not be a blocker for now. But if it is true, we may want to document this limitation.

File: emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java (hunk 2 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java:76]: Changing to error.

File: emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java (hunk 3 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java:73]: Changing to error.

File: emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java (hunk 1 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/main/java/com/amazon/aws/emr/mapping/MappingInvoker.java:55]: Formatted again.

File: emr-user-role-mapper-application/pom.xml (hunk 4 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/pom.xml:109]: Done.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestsUserService.java (hunk 6 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestsUserService.java:17]: I did not want to duplicate the same code. In actuality the `CommandBasedPrincipalResolver` could be static methods, and there is no encapsulation change by changing to public.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestBase.java (hunk 5 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestBase.java:23]: Done.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java (hunk 7 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java:117]: Done.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java (hunk 8 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java:174]: So what the test does is 
- mapping present, check we get credentials
- Upload a new mapping file where user has no role mapping
- Check that we get back empty credentials
Renamed the test to make the intention clearer.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java (hunk 10 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java:28]: The role and policies have user name added to them. The policy name has the hashcode appened which is a functioon of username - so I think multiple users can run the tests.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java (hunk 9 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java:53]: We create a new policy on each test (new ARN). So this case should not arise I think.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java (hunk 10 above)
Comment (yInnovation12) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java:28]: Yea.. multiple users can run tests in parallel but not same user I guess. It's ok for now.

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java (hunk 8 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/defaultmapper/DefaultMappingProviderImplIntegrationTest.java]: Changed the name of test.

**Review by yInnovation12 — APPROVED** [review]

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java (hunk 9 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/utils/IAMUtils.java:53]: It's idempotent.

```
aws iam attach-role-policy --policy-arn arn:aws:iam::176430881729:policy/urm-policy-107149120 --role-name test-integ-urm-role---871835695--463378646
a483e77e76c2:amazon-emr-user-role-mapper shkala$ aws iam attach-role-policy --policy-arn arn:aws:iam::176430881729:policy/urm-policy-107149120 --role-name test-integ-urm-role---871835695--463378646
```

File: emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestsUserService.java (hunk 6 above)
Comment (sharad-oss) [reply emr-user-role-mapper-application/src/test/com/amazon/aws/emr/integration/IntegrationTestsUserService.java:17]: As we talked to avoid code duplication the access was changed. The method by itself it like a static method with no encapsulation issues.
